File: 프로그래머스/lv3/42898. 등굣길/등굣길.java
class Solution {
    public int solution(int m, int n, int[][] puddles) {
        int answer = 0;
        
        long[][] map = new long[n][m];
		
		for(int i=0; i<puddles.length; i++) {
			int x = puddles[i][1]-1;
			int y = puddles[i][0]-1;
			map[x][y] = -1;
		}
		
		map[0][0] = 1;
		for(int i=0; i<n; i++) {
			for(int j=0; j<m; j++) {
				
				if(i == 0 && j == 0) continue;
				
				long value = 0;
				if(i-1 >=0 && map[i-1][j] != -1) {
					value += map[i-1][j] % 1000000007;
				}
				
				if(j-1 >=0 && map[i][j-1] != -1) {
					value += map[i][j-1] % 1000000007;
				}
				
				if(map[i][j] != -1) {
					map[i][j] = value;
				}
				
			}
		}
		
		answer = (int) (map[n-1][m-1] % 1000000007);
        
        return answer;
    }
}

File: 프로그래머스/3/42898. 등굣길/등굣길.py
def solution(m, n, puddles):
    answer = 0
    
    dp = [[0] * m for _ in range(n)]
    
    dp[0][0] = 1
    
    mod = 1000000007
    
    for puddle in puddles :
        x = puddle[1] - 1
        y = puddle[0] - 1
        dp[x][y] = -1
    
    for index in range(1, m):
        if dp[0][index] == -1:
            break
        
        dp[0][index] = 1
    
    for index in range(1, n):
        if dp[index][0] == -1:
            break
        dp[index][0] = 1
    
    for x in range(1, n):
        for y in range(1, m):
            if dp[x][y] == -1:
                continue
            
            if dp[x-1][y] >= 0:
                dp[x][y] = (dp[x][y] % mod + dp[x-1][y] % mod) % mod
            if dp[x][y-1] >= 0:
                dp[x][y] = (dp[x][y] % mod + dp[x][y-1] % mod) % mod
    
    answer = dp[n-1][m-1]
    
    
    return answer